internal/states/encryption/encryptionflow/flow.go

```diff
@@ -0,0 +1,123 @@
+package encryptionflow
+
+import (
+	"github.com/hashicorp/go-hclog"
+	"github.com/opentofu/opentofu/internal/states/encryption/encryptionconfig"
+)
+
+// EncryptionTopLevelJsonKey is the json key used to mark state or plans as encrypted.
+//
+// Changing this will invalidate all existing encrypted state, so migration code would have to be added
+// that can deal with the old key.
+//
+// Unencrypted state must not contain this key, so you cannot choose any of the top-level keys of the current
+// state data structure (currently stateV4).
+const EncryptionTopLevelJsonKey = "encryption"
+
+// New creates an encryption flow with the specified configuration.
+func New(
+	configKey encryptionconfig.Key,
+	encryptionConfig *encryptionconfig.Config,
+	fallbackConfig *encryptionconfig.Config,
+	logger hclog.Logger,
+) Flow {
+	return &flow{
+		configKey:        configKey,
+		logger:           logger,
+		encryptionConfig: encryptionConfig,
+		fallbackConfig:   fallbackConfig,
+	}
+}
+
+// Flow represents the top-level state or plan encryption/decryption flow
+// for a particular encryption configuration.
+//
+// You can obtain a copy of this interface by creating a Builder first and supplying it with the correct
+// encryption configuration, then calling the Build function.
+//
+// Note, that all encrypted data must be in JSON format and must contain a key specified in EncryptionTopLevelJsonKey
+// to indicate that it is valid encrypted data.
+type Flow interface {
+	StateFlow
+	PlanFlow
+}
+
+// StateFlow contains the encryption and decryption functions that should be used for state files only.
+// State files are JSON, so they are treated accordingly.
+type StateFlow interface {
+	// DecryptState decrypts encrypted state.
+	//
+	// You can pass it an encrypted or decrypted state in JSON format as a []byte. If the state is already decrypted
+	// (has no EncryptionTopLevelJsonKey), the function will return the state unmodified.
+	//
+	// If the state is encrypted, the method will attempt to decrypt the state according to its internal configuration,
+	// first with the primary encryption key, then with the fallback decryption key. If neither produces a valid result
+	// it fails.
+	DecryptState(payload []byte) ([]byte, error)
+
+	// EncryptState encrypts plaintext state.
+	//
+	// To encrypt the state, pass the unencrypted state in JSON format to this function. This function will attempt to
+	// encrypt it with the encryption key and return a valid, but encrypted JSON.
+	//
+	// If no encryption is configured, but "enforced" is enabled, the encryption will fail. Otherwise, this function
+	// will return the state unmodified.
```

Comment:
I think this statement regarding `"enforced"` is also correct for `EncryptPlan` and also for the decrypt methods. Is that correct?
If so, we should probably add this statement to the relative comments

Reply:
I believe the enforce flag is a config setting in the state backend or remote data source block. In other words, I don't think it would affect the plan, cause there's no place to set this enforcing for a plan.

Reply:
Then the current naming and structure is incorrect, we shouldn't be allowing the `enforced` value for configurations that are not for these purposes.

Reply:
I would not introduce multiple different structures. That'll make the code really complex.

It's just a case of an invalid configuration, which should lead to an error being returned by `Validate()`. 

There are lots of ways configurations can be invalid:
- setting key_provider.name to an unregistered key provider name
- setting a parameter under key_provider.config that does not exist
- not providing a required parameter under key_provider.config that is required by a key provider
- setting a parameter under key_provider.config to a value that a key provider does not allow
- setting method.name to an unregistered method name
- setting a parameter under method.confit that does not exist
- not providing a required parameter under method.config that is required by a method
- setting a parameter under method.config to a value that an encryption method does not allow
- setting enforced on a decryption fallback configuration
- setting enforced on an encryption configuration for plan files (encryption will fail for plan files anyway if they can't be encrypted)

Any of these needs to end the run with a non-zero exit code and cause an error to be printed for the user before any processing has taken place.

All that is important are good error messages that explain to the user what is wrong.

The final versions of the various validation functions will emerge when the flow, the methods, and the key providers are actually implemented.